### ISSUE TITLE
Selectable Province Map

### ISSUE THREAD

**Issue by EPrograming** [EPrograming/Project-Strategy#3]
Create a map with provinces which when selected are highlighted and a UI opens showing province info. 

Example map:
![provinces](https://user-images.githubusercontent.com/18630012/28946949-4ed3d2b0-787b-11e7-92b4-b5e84b0a608a.png)

**Comment by Pjaerr** [comment]
I saw your comment on my repo, do you need help with this? I got a system exactly like this working.

**Comment by EPrograming** [comment]
@Pjaerr Yes, i have a few provinces in but i am having trouble getting json to work well. You can fork the project, edit it then create a new pull request

**Comment by Pjaerr** [comment]
@EPrograming I'll take a look when I have some time.

**Comment by Pjaerr** [comment]
Hey @EPrograming, could you describe your problem in more detail so I can get a foundation within the project. Thanks!

**Comment by EPrograming** [comment]
I have a few provinces that are selectable, and open up ui, but i am not sure on how to implement importing province info from json for easear edditablity.

**Comment by EPrograming** [comment]
I also need to get around to making all the provinces, i only have the uk ones and 2 american ones

**Comment by Pjaerr** [comment]
Hey, I edited the code a little. I think trying to grab the json data from one big json file would cause too many problems, I think going with your initial idea of having each province be its own, small, json file is the best. I changed the code so that it will go through the Provinces folder, and for any json files, grab the data, storing it inside of a struct and then pushes that data into a new Province object stored inside a List. It will create a Province object for each json file inside of the folder and store that Province object in the provinces list where you can access the province by search for its json name such as if(provinces[i].name == "Ireland){return provinces[i]} < something like that.

I didn't want to fork or create a new branch just yet as I am a version behind you on Unity, and using different OS. Considering it is only 1 file, you could copy the code and mess around with it, see if it is what you wanted.

The code:

https://gist.github.com/Pjaerr/1b749d6da0167cb3b7fd30bf997ef5d4

**Comment by ProDevGuy** [comment]
@Pjaerr The code looks good. I will try it in a little bit. Btw unity works the same accross platforms. I regulary move between windows and linux and it works the same

**Comment by Pjaerr** [comment]
@ProDevGuy Oh good, Ill see about that if and when I contribute to something larger/multiple scripts etc. Worth nothing that on the code above, I don't make use of adjacencies[] from the json file as I had no idea what it is used for, or its datatype, but I do store it on the C# struct that holds the json as a string array, also, the string array 'files[]' stores an entire path, from the furthest point back it can go, might be worth just cutting off by the AssetsStreaming folder as that is what you need, although it doesn't break anything to not do so.

**Comment by EPrograming** [comment]
@Pjaerr I was thinking of using  adjacencies[] as a way for game to know what provinces border each other, but if your system of unity moving does not use it, i will remove it

**Comment by Pjaerr** [comment]
@EPrograming If can easily do it. I just didn't code it into the province object, just add it as a parameter.

**Comment by EPrograming** [comment]
@Pjaerr  Your script works and i can use it to get data from the json files, but i am having trobule getting the ui to change depending on province clicked. My current code is this: (not changed yet for your script)

` 
		
		void Update () {
			//theText = province.text;
			if(provinceName == "Ireland"){
				path = Application.streamingAssetsPath + "/Provinces/Ireland.json";
			}
			if(provinceName == "Wales"){
				path = Application.streamingAssetsPath + "/Provinces/Wales.json";
			}
			jsonString = File.ReadAllText(path);
			Province IRELAND = JsonUtility.FromJson<Province>(jsonString);
			provinceNameText = provinceNameText.GetComponent<TextMesh>();
			provinceNameTextUI = provinceNameTextUI.GetComponent<Text>();
			provinceNameText.text = IRELAND.name;
			provinceNameTextUI.text = IRELAND.name;
		
		
	}`

Though i am thinking of changing it to a system where it loads the provinces in start() not in update() but i am not sure on how to get it to work.

**Comment by Pjaerr** [comment]
@EPrograming Can't you just go to the provinces list in Provinces.cs, search for a province that has province.name == "Ireland", and then update a UI.Text element with that province's province.name value? Or am I misunderstanding?

**Comment by EPrograming** [comment]
@Pjaerr  Would i put the `prvoinceNameText = province.name` in the update function or the selectable script?

**Comment by Pjaerr** [comment]
@EPrograming I think you shouldn't update the UI every frame in the Update function anyway. Just update it when it changes. 

If you Text object was called provinceNameText and your Province object in the provinces list was called Ireland, you would write provinceNameText.text = Ireland.name;  

You could have a function on selectable that updates all of the UI information as described above, and then just call that once if a click is detected, making sure to put a check in place to see if UI is already open to avoid updating it loads of times for no reason.

**Comment by EPrograming** [comment]
@Pjaerr  So something like this?

`      foreach(provinces.Province str in provinces.provinces)
                    {
                       IRELAND = str.name;
                       provinceNameText = provinceNameText.GetComponent<TextMesh>();
                       provinceNameTextUI = provinceNameTextUI.GetComponent<Text>();
                       provinceNameText.text = IRELAND.name;
                       provinceNameTextUI.text = IRELAND.name;
                    }`

This is all in a onclick() function

**Comment by EPrograming** [comment]
@Pjaerr  I got the script to work somewhat, i keep getting the error:Assets/Scripts/Selectable.cs(59,29): error CS0118: `Selectable.provscript' is a `field' but a `type' was expected
I am not sure what that means.

**Comment by Pjaerr** [comment]
@EPrograming If you send me the Selectable.cs script here, Ill take a look.

**Comment by Pjaerr** [comment]
@EPrograming Also, if you reference me, sure you can. Keep in mind, all my images were hand drawn and so won't be accurate.

**Comment by EPrograming** [comment]
@Pjaerr  Thank you, and here is the script:
[Selectable.cs.zip](https://github.com/EPrograming/Project-Strategy/files/1346537/Selectable.cs.zip)

**Comment by EPrograming** [comment]
@Pjaerr  You could just download the files from the repo, the create a pull request becouse unity is cross-platform

**Comment by Pjaerr** [comment]
@EPrograming because I am not working on anything large, just trying things, I don't want to go through that hassle.

I changed the code, this is how I'd set the UI text to a province's name. (With comments)

https://gist.github.com/Pjaerr/bb75b1664f1ae7a2a27175c81282d11f

**Comment by EPrograming** [comment]
@Pjaerr Your script does not work :( It broke allot of stuff:(

**Comment by Pjaerr** [comment]
@EPrograming I didn't test it in Unity, it should be clear what has changed and what has broken it. I can take a longer look and fork the project when I have some spare time.

**Comment by EPrograming** [comment]
@Pjaerr This is strange. It is printing stuff to console that i did not tell to print. All but 3 Debug.Logs are commented out and it is printing allot of different stuff. More that 3

**Comment by EPrograming** [comment]
@Pjaerr  this is after i reverted to the old script.

**Comment by Pjaerr** [comment]
If you describe exactly what you are wanting to do, maybe tomorrow or in a few days when I have sometime Ill fork the entire project and fix the issue with full comments. @EPrograming

**Comment by Pjaerr** [comment]
Just a quick addition. The solution for me, when making provinces clickable, is to store a reference to that provinces transform from the heirarchy inside of the PRovince object itself, and so every province also has a real-world object as well as its names and stats. @EPrograming

**Comment by EPrograming** [comment]
@Pjaerr  Would that be in the script or the json file>

**Comment by EPrograming** [comment]
@Pjaerr Goals :
High to Low Priorety
- Name
- Expandible Selectable province system
- Provinces for entire world
- UI

Long Term goals:

- AI
- Playable World
- Time
- Units
-

**Comment by Pjaerr** [comment]
Hey @EPrograming Can you explain here what you want to do with the Selectable.cs script? I have forked and cloned the project and can try and implement what you want doing.

**Comment by EPrograming** [comment]
@Pjaerr  what i want doing is when you click on the province, it opens up a ui showing the province name, owner and other values from the json file. I also want the provinces to be "glowing" when they are clicked

**Comment by Pjaerr** [comment]
Okay, on Monday Ill see what I can do in terms of the provinces opening their relative UI panels. lll just do a pull request when I have it done. @EPrograming

**Comment by EPrograming** [comment]
@Pjaerr  Provinces opening up ui panels allready exists, but they dont show the correct information

**Comment by Pjaerr** [comment]
Yeah, I said relative, so opening the panels that actually relate to them. Not just a default panel. @EPrograming

**Comment by EPrograming** [comment]
@Pjaerr Not related to this, but EU parlement is planning new copyright rules which threatens Open-Source code platforms like Github, Gitlab and others. 

Sign the petition! https://savecodeshare.eu/

**Comment by EPrograming** [comment]
@Pjaerr  For some reason, whenever i change the selectable.cs script, everything stops workiing

**Comment by EPrograming** [comment]
@Pjaerr   I moved the text changing code to the selectable.cs file,but for some reasion the ui says  the provinces selected is allways ireland.

**Comment by EPrograming** [comment]
@Pjaerr  Update: The script works better now, but when i click on ireland the text sais ireland for a second, then it goes back to wales. Wales works fine btw

**Comment by EPrograming** [comment]
@Pjaerr  Update 2: Everything works! I just need to create all the provinces. Also, i have a **_Very_** Inefficient way of doing it. I currently have an if statement for every province, so if you have a better way, please let me know
